feat: add canary deployment logic

mfrancisc opened the pull request:
This PR adds custom canary deployment logic which will redirect some users to our new UI at `sandbox.redhat.com`.

After merging this change, users visiting the first time the UI will get assigned with a random number from 0 to 100, the number will be stored in localstorage as `dev-sandbox.ui-canary-weight` and reused every time the user comes back to the UI (unless they do it with a different browser or they delete the localstorage value).

On the registration service backend we now have a new api `GET /api/v1/uiconfig` which returns a json 
`{ "uiCanaryDeploymentWeight": number}`. 

The `uiCanaryDeploymentWeight` indicates the threshold weight ,  thus users getting a number lower than that will be redirected to the new ui at `sandbox.redhat.com`, the others will stay in the current HCC UI.

This way the threshold can be configured in the backend, so that a bigger/smaller number of users can be routed to the new UI, which allows us to "force" users to go to the new UI and perform the rollout incrementally.

CodyWMitchell commented:
I believe Unleash feature-flags are typically used to roll out gradual changes. Is there a reason it's being done differently here?

mfrancisc commented:
@CodyWMitchell if I'm well understanding your question , in this case we are rolling out a complete new UI application that is hosted at `sandbox.redhat.com` , so I'm not sure we can use a feature flag, as this is not just a new feature in the sandbox HCC code base, but users will get redirected to a new domain/app.

florkbr commented on src/routes/SandboxPage/SandboxPage.tsx:
```diff
@@ -1,21 +1,52 @@
+
+  (async () => {
+    try {
+      await loadUIConfig();
```
@mfrancisc won't this be called every re-render of the `SandboxPage`? Can we use a relevant react hook to accomplish this and prevent unnecessary calls during renders?

mfrancisc replied:
this needs to be executed before loading the component . I mean we first need to see if the user should stay on this UI or not, and than return the component. Would work in the same way with the hook ?

mfrancisc replied:
I've implemented the hook, and moved it to the Routing component, as that is more appropriate since it needs to redirect before displaying the sandbox page.

florkbr commented on src/routes/SandboxPage/SandboxPage.tsx:
```diff
@@ -1,21 +1,52 @@
+   The threshold can be configured in the backend so that a bigger/smaller number of users can be routed to the new UI.
+   ***/
+  const loadUIConfig = async function () {
+    console.log('loading ui config');
```
@mfrancisc I'd suggest either removing these log statements - or placing them behind a check so they don't show up in prod envs.

mfrancisc replied:
removed

florkbr commented on src/utils/canary-deployment.ts:
```diff
@@ -0,0 +1,34 @@
+
+export const randomIntFromInterval = (min: number, max: number): number => {
+  // min and max included
+  return Math.floor(Math.random() * (max - min + 1) + min);
```
`Math.random()` does not provide cryptographically secure random numbers FWIW. For that we should use `Crypto.getRandomValues()` - though in this context it seems the random number is just used to random select users to navigate to the new UI.

mfrancisc replied:
Thanks for letting me know. But I think using `Math.random()` is enough here for our usecase.

florkbr reviewed (changes requested):
@mfrancisc we should be using feature flags via `Unleash` for this IMO. Unleash allows you to specify conditional rollouts of feature flags to a certain percentage of users.

So you could have `developer-sandbox.ui-redirect` feature toggle conditionally set to true for `%50` of users (on stage or prod). Then in the UI you'd use the `useFlag` hook from unleash: https://docs.getunleash.io/reference/sdks/react#check-feature-toggle-status

florkbr commented on src/utils/canary-deployment.ts:
```diff
@@ -0,0 +1,34 @@
+
+export const redirectUser = (userWeightValue: number, weightThreshold: number) => {
+  if (userWeightValue <= weightThreshold) {
+    console.log('user will be redirected to new UI');
```
More console.log's we should either remove or gate for prod envs in this file.

mfrancisc commented on src/routes/SandboxPage/SandboxPage.tsx:
```diff
@@ -1,21 +1,52 @@
+    const { getUIConfigData } = useRegistrationService();
+    const uiConfigData = await getUIConfigData();
+    if (uiConfigData == undefined) {
+      console.log('Unable to load uiconfig. Got undefined response.');
```
no we don't need to show anything to the user. I think it's enough to log this in the console for debugging. I can turn it in `console.error` indeed.

mfrancisc replied:
done

mfrancisc commented:
@florkbr I've addressed your comments. 

As per the Unleash implementation, we can consider that for later since, as discussed in slack:
1. We'll need to onboard the team to app interface for having permissions to access and configure using the Unleash dashboard
2. We'll need to understand how to implement exactly the same functionality using Unleash
3. This is just a temporary solution, we will not need it in a couple of weeks

If there's no others concerns with the current code, I will move forward with the current implementation and we can improve it in follow up pr's.

mfrancisc commented:
It also looks like the konflux pipeline failed : https://konflux-ui.apps.stone-prd-rh01.pg1f.p1.openshiftapps.com/ns/rhtap-releng-tenant/applications/my-app-1001/pipelineruns/devfile-sample-code-with-quarkus-on-pull-request-rzq76 , and I don't see any relevant logs and cannot re-trigger it ( lacking permissions apparently ).  Not sure what the problem is.

@ryelo @florkbr @Hyperkid123 could you guys please rerun it or let me know what is the problem ? 🙏
